src/Npgsql/Internal/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -157,77 +264,172 @@ public override int ValidateAndGetLength(decimal value, NpgsqlParameter? paramet
 
-            while (groupCount > 0)
-                buf.WriteInt16(groups[--groupCount]);
+                while (groupCount > 0)
+                    buf.WriteInt16(groups[--groupCount]);
+            }
         }
 
         /// <inheritdoc />
-        public void Write(short value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)  => Write((decimal)value, buf, parameter);
+        public Task Write(short value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(int value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)    => Write((decimal)value, buf, parameter);
+        public Task Write(int value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(long value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)   => Write((decimal)value, buf, parameter);
+        public Task Write(long value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(byte value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)   => Write((decimal)value, buf, parameter);
+        public Task Write(byte value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(float value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)  => Write((decimal)value, buf, parameter);
+        public Task Write(float value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
         /// <inheritdoc />
-        public void Write(double value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter) => Write((decimal)value, buf, parameter);
+        public Task Write(double value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async, CancellationToken cancellationToken = default)
+            => Write((decimal)value, buf, lengthCache, parameter, async, cancellationToken);
+
+        static ushort[] FromBigInteger(BigInteger value)
+        {
+            var str = value.ToString(CultureInfo.InvariantCulture);
```

Comment:
The built in ToString() is O(n^2), where n is the number of bits in the number, but a quite efficient such a one. Note that it can use the internal uint[] array directly. If we use ToByteArray() it must then be converted to a byte array, and then we would probably want to ourselves convert it back to uint[]. I managed to write a custom implementation that is like 3x faster for the absolutely largest numbers (similar to the method used while reading, but switching the bases 2^32 and 10^4), but not that much faster for small-sized to mid-sized numbers. Since we must then do everything in base 10 (or a power of 10), I had to implement Karatsuba multiplication directly etc. to get good performance. But for at least an initial implementation, I don't think it's worth to bloat the code with this.

The reading is quite a different story, the BigInteger.Parse method is insanely slow, and my alternative implementation is quite small and can make use of the BigInteger class to do the calculations.

Reply:
Sounds convincing to me :) You're the expert...